lib/controllers/history_controller.dart

```diff
@@ -0,0 +1,113 @@
+import 'dart:collection';
+
+import 'package:flutter/material.dart';
+import 'package:history_manager/history_manager.dart';
+
+/// - Holds undo and redo functionalities for the session of this controller.
+/// - Extends to [ChangeNotifier] to notify changes for the mutable properties to take changes in the UI State.
+class HistoryController extends ChangeNotifier {
+  HistoryController();
+
+  // * Mutable Properties
+  /// Stack of executed changes during a session.
+  Queue<ExecutedChange> _executedChanges = ListQueue();
+
+  /// Stack of executed undos and that can be redo upon undoing a specific action.
+  Queue<ExecutedChange> _redos = ListQueue();
+
+  // * Getters
+  /// Checker if stack can still redo actions
+  bool get canRedo => _redos.isNotEmpty;
+
+  /// Checker if stack can still undo actions
+  bool get canUndo => _executedChanges.isNotEmpty;
+
+  /// Returns the value _executedChanges.
+  Queue<ExecutedChange> get executedChanges => _executedChanges;
+
+  /// Returns the value _redos.
+  Queue<ExecutedChange> get redos => _redos;
+
+  // * Setters
+  /// Sets value for `_redos`
+  set redos(Queue<ExecutedChange> redo) {
+    _redos = redo;
+    notifyListeners();
+  }
+
+  /// Sets value for `_executedChanges`
+  set executedChanges(Queue<ExecutedChange> executedChanges) {
+    _executedChanges = executedChanges;
+    notifyListeners();
+  }
+
+  // * Public methods
+
+  /// - Redo the last undone action.
+  /// - Checks if `_redos` stack is not empty to be able to know if there is an action that cane be redone.
+  /// - Calls all `execute()` method of each change to apply the changes for the session.
+  /// - Adds the redoable change to `_executedChanges`.
+  void redo() {
+    if (canRedo) {
+      final change = _redos.removeFirst();
+      change.execute();
+      // changes.forEach((change) => change.execute());
+      _executedChanges.addLast(change);
+      notifyListeners();
+    }
+  }
+
+  /// - Undo the last change from the session.
+  /// - Checks if `_executedChanges` is not empty to be able to undone the last change.
+  /// - Adds the change at the first index of `_redos` stack.
+  void undo() {
+    if (canUndo) {
+      final change = _executedChanges.removeLast();
+      // changes.forEach((change) => change.undo());
+      change.undo();
+      _redos.addFirst(change);
```

Comment:
See above. You should still be able to achieve similar logic using add and removeLast only.